File: src/csharp/Foundgine.Core/Semantic/Planning/Mutation/MutationAuthorizer.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Authorization;
using Foundgine.Core.Semantic.Mutation;
using Foundgine.Core.Semantic.Query;

namespace Foundgine.Core.Semantic.Planning.Mutation;

/// <summary>
/// Applies semantic write authorization to a provider-independent mutation
/// plan. It deliberately sits outside MutationPlanner so planning remains a
/// structural concern while policy remains a semantic concern.
/// </summary>
public sealed class MutationAuthorizer
{
    private readonly IMutationSchema _schema;
    private readonly ISemanticAuthorizationPolicy _policy;

    public MutationAuthorizer(IMutationSchema schema, ISemanticAuthorizationPolicy policy)
    {
        _schema = schema ?? throw new ArgumentNullException(nameof(schema));
        _policy = policy ?? throw new ArgumentNullException(nameof(policy));
    }

    public MutationPlan Authorize(MutationPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        foreach (var operation in plan.Operations)
            Authorize(operation);
        return plan;
    }


    /// <summary>
    /// Authorizes the canonical semantic mutation plan directly. The authorized
    /// semantic representation is therefore the same representation that is
    /// subsequently lowered for execution; no independently reconstructed batch
    /// can become the execution source of truth.
    /// </summary>
    public SemanticMutationPlan Authorize(SemanticMutationPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);

        foreach (var operation in plan.Operations)
        {
            var entity = _schema.GetEntity(operation.Entity);
            RequireAllowed(
                _policy.GetEntityAccess(entity.Id, AuthorizationOperation.Write),
                $"write entity '{entity.Name}'");

            foreach (var field in operation.Fields)
            {
                if (!entity.Fields.ContainsKey(field.Field))
                    throw new InvalidOperationException(
                        $"Mutation field '{field.Field.Value}' is not registered on '{entity.Name}'.");

                RequireAllowed(
                    _policy.GetFieldAccess(entity.Id, field.Field, AuthorizationOperation.Write),
                    $"write field '{entity.Name}.{field.Field.Value}'");
            }

            foreach (var field in operation.ConflictFields)
            {
                if (!entity.Fields.ContainsKey(field))
                    throw new InvalidOperationException(
                        $"Conflict field '{field.Value}' is not registered on '{entity.Name}'.");

                RequireAllowed(
                    _policy.GetFieldAccess(entity.Id, field, AuthorizationOperation.Write),
                    $"write conflict field '{entity.Name}.{field.Value}'");
            }

            foreach (var field in operation.ReturnFields)
            {
                if (!entity.Fields.ContainsKey(field))
                    throw new InvalidOperationException(
                        $"Return field '{field.Value}' is not registered on '{entity.Name}'.");

                RequireAllowed(
                    _policy.GetFieldAccess(entity.Id, field, AuthorizationOperation.Read),
                    $"read return field '{entity.Name}.{field.Value}'");
            }

            ValidateFilter(operation.Filter, entity);
        }

        return plan;
    }

    public MutationBatchPlan Authorize(MutationBatchPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        foreach (var operation in plan.Operations)
            Authorize(operation);
        return plan;
    }

    private void Authorize(MutationOperation operation)
    {
        var entity = operation.Entity;
        RequireAllowed(
            _policy.GetEntityAccess(entity.Id, AuthorizationOperation.Write),
            $"write entity '{entity.Name}'");

        foreach (var field in operation.Fields)
        {
            var fieldMapping = entity.Fields.FirstOrDefault(pair => pair.Value == field.Column);
            if (fieldMapping.Equals(default(KeyValuePair<FieldId, ColumnId?>)))
                throw new InvalidOperationException(
                    $"Mutation field column '{field.Column.Value}' has no semantic field mapping on '{entity.Name}'.");

            var fieldId = fieldMapping.Key;

            RequireAllowed(
                _policy.GetFieldAccess(entity.Id, fieldId, AuthorizationOperation.Write),
                $"write field '{entity.Name}.{fieldId.Value}'");
        }

        foreach (var fieldId in operation.ReturnFields ?? Array.Empty<FieldId>())
        {
            RequireAllowed(
                _policy.GetFieldAccess(entity.Id, fieldId, AuthorizationOperation.Read),
                $"read return field '{entity.Name}.{fieldId.Value}'");
        }

        ValidateFilter(operation.Filter, entity);
    }

    private void ValidateFilter(SemanticFilterExpression? filter, MutationEntitySchema entity)
    {
        if (filter is null)
            return;

        switch (filter)
        {
            case SemanticFieldFilter field:
                RequireAllowed(
                    _policy.GetFieldAccess(entity.Id, field.Field, AuthorizationOperation.Read),
                    $"filter on field '{entity.Name}.{field.Field.Value}'");
                break;

            case SemanticRelationshipFilter relationship:
                RequireAllowed(
                    _policy.GetRelationshipAccess(entity.Id, relationship.Relationship, AuthorizationOperation.Read),
                    $"filter through relationship '{relationship.Relationship.Value}'");
                var relationshipSchema = _schema.GetRelationship(relationship.Relationship);
                ValidateFilter(relationship.Predicate, _schema.GetEntity(relationshipSchema.Target));
                break;

            case SemanticAggregateFilter aggregate:
                RequireAllowed(
                    _policy.GetRelationshipAccess(entity.Id, aggregate.Relationship, AuthorizationOperation.Read),
                    $"aggregate filter through relationship '{aggregate.Relationship.Value}'");
                if (aggregate.Field is { } aggregateField)
                {
                    var aggregateRelationshipSchema = _schema.GetRelationship(aggregate.Relationship);
                    RequireAllowed(
                        _policy.GetFieldAccess(aggregateRelationshipSchema.Target, aggregateField,
                            AuthorizationOperation.Read),
                        $"aggregate filter field '{aggregateField.Value}'");
                }

                break;

            case SemanticAndFilter and:
                foreach (var expression in and.Expressions)
                    ValidateFilter(expression, entity);
                break;

            case SemanticOrFilter or:
                foreach (var expression in or.Expressions)
                    ValidateFilter(expression, entity);
                break;

            default:
                throw new NotSupportedException(
                    $"Mutation authorization does not support filter '{filter.GetType().Name}'.");
        }
    }

    private static void RequireAllowed(AuthorizationDecision decision, string resource)
    {
        if (!decision.IsAllowed)
            throw new SemanticAuthorizationException($"Access denied for {resource}.");
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/mutation/MutationAuthorizer.java
package com.foundgine.core.semantic.planning.mutation;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.authorization.*;
import com.foundgine.core.semantic.mutation.SemanticMutationPlan;
import com.foundgine.core.semantic.query.*;
import java.util.*;

/** Applies semantic write/read authorization to provider-independent mutation plans. */
public final class MutationAuthorizer {
    private final MutationSchema schema; private final ISemanticAuthorizationPolicy policy;
    public MutationAuthorizer(MutationSchema schema, ISemanticAuthorizationPolicy policy){this.schema=Objects.requireNonNull(schema);this.policy=Objects.requireNonNull(policy);}

    public MutationPlan authorize(MutationPlan plan){Objects.requireNonNull(plan);plan.operations().forEach(this::authorize);return plan;}
    public MutationBatchPlan authorize(MutationBatchPlan plan){Objects.requireNonNull(plan);plan.operations().forEach(this::authorize);return plan;}
    public SemanticMutationPlan authorize(SemanticMutationPlan plan){
        Objects.requireNonNull(plan); for(var op:plan.operations()){var entity=schema.getEntity(op.entity());require(policy.getEntityAccess(entity.id(),AuthorizationOperation.WRITE),"write entity '"+entity.name()+"'");
            for(var field:op.fields()){if(!entity.fields().containsKey(field.field()))throw new IllegalStateException("Mutation field '"+field.field().value()+"' is not registered on '"+entity.name()+"'.");require(policy.getFieldAccess(entity.id(),field.field(),AuthorizationOperation.WRITE),"write field '"+entity.name()+"."+field.field().value()+"'");}
            for(var field:op.conflictFields()){if(!entity.fields().containsKey(field))throw new IllegalStateException("Conflict field '"+field.value()+"' is not registered on '"+entity.name()+"'.");require(policy.getFieldAccess(entity.id(),field,AuthorizationOperation.WRITE),"write conflict field '"+entity.name()+"."+field.value()+"'");}
            for(var field:op.returnFields()){if(!entity.fields().containsKey(field))throw new IllegalStateException("Return field '"+field.value()+"' is not registered on '"+entity.name()+"'.");require(policy.getFieldAccess(entity.id(),field,AuthorizationOperation.READ),"read return field '"+entity.name()+"."+field.value()+"'");}
            validateFilter(op.filter(),entity);
        } return plan;
    }
    private void authorize(MutationOperation op){var entity=op.entity();require(policy.getEntityAccess(entity.id(),AuthorizationOperation.WRITE),"write entity '"+entity.name()+"'");
        for(var field:op.fields()){var fieldId=entity.fields().entrySet().stream().filter(e->Objects.equals(e.getValue(),field.column())).map(Map.Entry::getKey).findFirst().orElseThrow(()->new IllegalStateException("Mutation field column '"+field.column().value()+"' has no semantic field mapping on '"+entity.name()+"'."));require(policy.getFieldAccess(entity.id(),fieldId,AuthorizationOperation.WRITE),"write field '"+entity.name()+"."+fieldId.value()+"'");}
        if(op.returnFields()!=null)for(var field:op.returnFields())require(policy.getFieldAccess(entity.id(),field,AuthorizationOperation.READ),"read return field '"+entity.name()+"."+field.value()+"'");validateFilter(op.filter(),entity);
    }
    private void validateFilter(SemanticFilterExpression filter,MutationEntitySchema entity){if(filter==null)return;
        if(filter instanceof SemanticFieldFilter f){require(policy.getFieldAccess(entity.id(),f.field(),AuthorizationOperation.READ),"filter on field '"+entity.name()+"."+f.field().value()+"'");return;}
        if(filter instanceof SemanticRelationshipFilter r){require(policy.getRelationshipAccess(entity.id(),r.relationship(),AuthorizationOperation.READ),"filter through relationship '"+r.relationship().value()+"'");var rel=schema.getRelationship(r.relationship());validateFilter(r.predicate(),schema.getEntity(rel.target()));return;}
        if(filter instanceof SemanticAggregateFilter a){require(policy.getRelationshipAccess(entity.id(),a.relationship(),AuthorizationOperation.READ),"aggregate filter through relationship '"+a.relationship().value()+"'");if(a.field()!=null){var rel=schema.getRelationship(a.relationship());require(policy.getFieldAccess(rel.target(),a.field(),AuthorizationOperation.READ),"aggregate filter field '"+a.field().value()+"'");}return;}
        if(filter instanceof SemanticAndFilter a){a.expressions().forEach(x->validateFilter(x,entity));return;}
        if(filter instanceof SemanticOrFilter o){o.expressions().forEach(x->validateFilter(x,entity));return;}
        throw new UnsupportedOperationException("Mutation authorization does not support filter '"+filter.getClass().getSimpleName()+"'.");
    }
    private static void require(AuthorizationDecision decision,String resource){if(!decision.isAllowed())throw new SemanticAuthorizationException("Access denied for "+resource+".");}
}
